State API requires eTags (should be optional)

**Version**
Latest from composer as 22nd Jan 2015

**Steps to reproduce the bug**
1. Store an application/json document in the State API. 
2. Update the document (PUT) omitting the etag header. 

**Expected behaviour**
Document should be accepted, returning 204 no content. Etags are not required for the State API, only the Agent and Activity Profile APIs. 

**Actual behaviour**
400 Bad request is returned. 

**Server information**
MongoDB with Basic auth. Running on MAMP. 

**Client information**
OS: OS X 10.10.1
Browser: Chrome

**Additional information**
I've tested the client code I'm using against Wax and Cloud and it works.

Conformance test output:
```
  4265 passing (10m)
  242 pending
  5 failing

  1) State concurrency Good state concurrency: correctMatchWithRegistration request cluster Then the stateConcurrency response is verified:

      Uncaught AssertionError: "9404f3e9c267c819ec754de3e853b4dd4dfde33e" == "102b16ce866bbbb6cea75cb663dcdcb0712100a3"
      + expected - actual

      +102b16ce866bbbb6cea75cb663dcdcb0712100a3
      -9404f3e9c267c819ec754de3e853b4dd4dfde33e

      at Object.obj.validate (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\fixtures\clusters\stateConcurrency.js:40:40)
      at Object.<anonymous> (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\steps\verify.js:119:43)
      at Request._callback (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\utils\request.js:239:13)
      at Request.self.callback (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\request\request.js:123:22)
      at Request.emit (events.js:98:17)
      at Request.<anonymous> (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\request\request.js:1047:14)
      at Request.emit (events.js:117:20)
      at IncomingMessage.<anonymous> (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\request\request.js:998:12)
      at IncomingMessage.emit (events.js:117:20)
      at _stream_readable.js:943:16

  2) State concurrency Bad state concurrency: incorrectNoneMatch request cluster Then the LRS responds with HTTP 412:
     Uncaught AssertionError: Response status: 204, expected: [412]
      at Object.module.exports.statusCodes (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\utils\assertions.js:9:16)
      at Object.library.then.main (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\steps\base.js:283:20)
      at Object.invoke (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\fn.js:32:19)
      at interpret (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Macro.js:50:12)
      at interpret_step (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Interpreter.js:71:47)
      at iterator (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Interpreter.js:63:19)
      at iterate (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Array.js:72:13)
      at Array.each_async (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Array.js:78:9)
      at Array.each_async (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\fn.js:27:23)
      at interpret (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Interpreter.js:58:21)

  3) State concurrency Bad state concurrency: incorrectMatch request cluster Then the LRS responds with HTTP 412:
     Uncaught AssertionError: Response status: 204, expected: [412]
      at Object.module.exports.statusCodes (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\utils\assertions.js:9:16)
      at Object.library.then.main (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\steps\base.js:283:20)
      at Object.invoke (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\fn.js:32:19)
      at interpret (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Macro.js:50:12)
      at interpret_step (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Interpreter.js:71:47)
      at iterator (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Interpreter.js:63:19)
      at iterate (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Array.js:72:13)
      at Array.each_async (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Array.js:78:9)
      at Array.each_async (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\fn.js:27:23)
      at interpret (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Interpreter.js:58:21)

  4) State concurrency Bad state concurrency: ifMatchAndIfNoneMatch request cluster Then the LRS responds with HTTP 412:
     Uncaught AssertionError: Response status: 204, expected: [412]
      at Object.module.exports.statusCodes (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\utils\assertions.js:9:16)
      at Object.library.then.main (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\steps\base.js:283:20)
      at Object.invoke (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\fn.js:32:19)
      at interpret (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Macro.js:50:12)
      at interpret_step (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Interpreter.js:71:47)
      at iterator (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Interpreter.js:63:19)
      at iterate (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Array.js:72:13)
      at Array.each_async (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Array.js:78:9)
      at Array.each_async (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\fn.js:27:23)
      at interpret (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Interpreter.js:58:21)

  5) statement structure id test Bad id: '7396683c-0759-e411-b974-a0d3c123d81f' ("e" in 3rd segment not acceptable version) Then the LRS responds with
 HTTP 200:
     AssertionError: Response status: 409, expected: [200]
      at Object.module.exports.statusCodes (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\utils\assertions.js:9:16)
      at Object.library.then.main (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\steps\base.js:283:20)
      at Object.invoke (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\fn.js:32:19)
      at interpret (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Macro.js:50:12)
      at interpret_step (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Interpreter.js:71:47)
      at iterator (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Interpreter.js:63:19)
      at iterate (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Array.js:72:13)
      at Array.each_async (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Array.js:78:9)
      at Array.each_async (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\fn.js:27:23)
      at interpret (Z:\git\forks\ryansmith94\xAPI_LRS_Test-ll\src\node_modules\yadda\lib\Interpreter.js:58:21)



Warning: Task "mochaTest:stage1-core" failed. Use --force to continue.
```

--- a/app/locker/repository/Document/EloquentDocumentRepository.php
+++ b/app/locker/repository/Document/EloquentDocumentRepository.php
@@ -312,7 +312,7 @@ class EloquentDocumentRepository implements DocumentRepository {
   }
 
   private function checkETag($sha, $ifMatch, $ifNoneMatch, $noConflict = true) {
-    $ifMatch = isset($ifMatch) ? '"'.strtoupper($ifMatch).'"' : null;
+    $ifMatch = isset($ifMatch) ? strtoupper($ifMatch) : null;
 
     if (isset($ifMatch) && $ifMatch !== $sha) {
       throw new FailedPrecondition('Precondition (If-Match) failed.'); // 412.

--- a/app/locker/repository/Document/EloquentDocumentRepository.php
+++ b/app/locker/repository/Document/EloquentDocumentRepository.php
@@ -226,13 +226,6 @@ class EloquentDocumentRepository implements DocumentRepository {
 
     $existing_document = $this->findStateDoc( $lrs, $data['stateId'], $data['activityId'], $data['agent'], $data['registration'], true );
 
-    if ($method === 'PUT') $this->checkETag(
-      isset($existing_document->sha) ? $existing_document->sha : null,
-      $data['ifMatch'],
-      $data['ifNoneMatch'],
-      false
-    );
-
     if( !$existing_document ){
       $document                 = $this->documentapi;
 

--- a/app/models/DocumentAPI.php
+++ b/app/models/DocumentAPI.php
@@ -15,7 +15,7 @@ class DocumentAPI extends Eloquent {
   /**
    * Returns true if $array is associative.
    * @param Array $array
-   * @return boolean      
+   * @return boolean
    */
   private function isJSON($array) {
     return is_array($array) && array_keys($array) !== range(0, count($array) - 1);
@@ -24,8 +24,9 @@ class DocumentAPI extends Eloquent {
   private function putContent($content, $contentType) {
     switch ($contentType) {
       case 'application/json':
-        $this->setSha($this->content);
-        $this->overwriteContent(json_decode($content, true));
+        $encoded_content = gettype($this->content) === 'string' ? $this->content : json_encode($this->content);
+        $this->setSha($encoded_content);
+        $this->overwriteContent(json_decode($encoded_content, true));
         break;
       case 'text/plain':
         $this->setSha($content);
@@ -94,13 +95,13 @@ class DocumentAPI extends Eloquent {
       $content->move($dir, $filename);
     } else {
       $ext = array_search($contentType, FileTypes::getMap());
-      
+
       $filename = time().'_'.mt_rand(0,1000).($ext !== false ? '.'.$ext : '');
 
       $size = file_put_contents($dir.$filename, $content);
 
       if ($size === false) throw new \Exception('There was an issue saving the content');
-    } 
+    }
 
     $this->content = $filename;
     $this->setSha($content);

